lib/documents.js

```diff
@@ -2263,12 +2268,184 @@ function writeAllDocumentsImpl(options) {
-  let result = requester.startRequest(operation);
-  operation.response = result;
+
+  let jobState = {
+    docInstance : this,
+    stream: inputStream,
+    requesterCount: 0,
+    docsWrittenSuccessfully:0,
+    docsFailedToBeWritten: 0,
+    jobOptions: (jobOptions)? mlutil.copyProperties(jobOptions):{}
+  };
+
+ if(jobState.jobOptions){
+   // Check for jobOptions.concurrentRequests
+   if(jobState.jobOptions.concurrentRequests) {
+     const concurrentRequests = jobState.jobOptions.concurrentRequests;
+     if (concurrentRequests.multipleOf) {
+       if (!new Set(['forests', 'hosts']).has(String(concurrentRequests.multipleOf))) {
+         throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+       }
+     } else {
+       concurrentRequests.multipleOf = 'forests';
+     }
+
+     if (concurrentRequests.multiplier && concurrentRequests.multiplier < 0) {
+       throw new Error('concurrentRequests.multiplier cannot be less than zero');
+     } else {
+       concurrentRequests.multiplier = 4;
+     }
+     jobState.jobOptions.concurrentRequests = concurrentRequests;
+   }
+   else {
+     jobState.jobOptions.concurrentRequests = {multipleOf:'forests', multiplier:4};
+   }
+
+   // Check for jobOptions.batchSize
+   jobState.jobOptions.batchSize = (jobState.jobOptions.batchSize && jobState.jobOptions.batchSize >0) ? jobState.jobOptions.batchSize:100;
+
+   //Check for defaultMetadata
+   jobState.jobOptions.defaultMetadata = (jobState.jobOptions.defaultMetadata)? jobState.jobOptions.defaultMetadata:{};
+
+   //Check for transform
+   jobState.jobOptions.transform = (jobState.jobOptions.transform)? jobState.jobOptions.transform:{};
+ }
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onWriteAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
   return inputStream;
 }
 
+function finishWriter(jobState) {
+  jobState.requesterCount--;
+  if(jobState.requesterCount === 0) {
+    if(jobState.jobOptions && jobState.jobOptions.onCompletion){
+      const summary = {
+        docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+        docsFailedToBeWritten: jobState.docsFailedToBeWritten,
+        timeElapsed: jobState.timeElapsed
+      };
+      if (jobState.error) {
+        summary.error = jobState.error.toString();
+      }
+      jobState.jobOptions.onCompletion(summary);
+    }
+  }
+}
+
+function onWriteAllInit(output) {
+
+  let maxRequesters = 0;
+  const jobState = this;
+
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'forests':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
```

Comment:
Because this is a deferred callback, an error thrown here won't be received by the application.

Instead, it would be better to emit an error on the stream.